File: platforms/react-native/ios/RaTeXView.swift
// RaTeXView.swift — UIKit view and SwiftUI wrapper for rendering a LaTeX formula.

import UIKit
import SwiftUI

// MARK: - UIKit

/// A UIView that renders a LaTeX formula using the RaTeX engine.
///
/// ```swift
/// let view = RaTeXView()
/// view.latex = #"\frac{-b \pm \sqrt{b^2-4ac}}{2a}"#
/// view.fontSize = 28
/// ```
@MainActor
public class RaTeXView: UIView {

    // MARK: Public properties

    /// The LaTeX math-mode string to render.
    public var latex: String = "" {
        didSet { guard latex != oldValue else { return }; rerender() }
    }

    /// Font size in points. Determines the physical size of the formula.
    public var fontSize: CGFloat = 24 {
        didSet { guard fontSize != oldValue else { return }; rerender() }
    }

    /// Rendering mode. `true` (default) for display/block style (`$$...$$`);
    /// `false` for inline/text style (`$...$`).
    public var displayMode: Bool = true {
        didSet { guard displayMode != oldValue else { return }; rerender() }
    }

    /// Default formula color. Explicit LaTeX colors still take precedence.
    public var color: UIColor = .black {
        didSet { guard !color.isEqual(oldValue) else { return }; rerender() }
    }

    /// Called when a render error occurs (e.g. invalid LaTeX).
    public var onError: ((Error) -> Void)?

    // MARK: Private state

    private var renderer: RaTeXRenderer?

    // MARK: Init

    public override init(frame: CGRect) {
        super.init(frame: frame)
        backgroundColor = .clear
        // Redraw whenever the frame changes (needed in Fabric / any layout system
        // that sets the frame after updateProps has already called setNeedsDisplay).
        contentMode = .redraw
    }

    public required init?(coder: NSCoder) {
        super.init(coder: coder)
        backgroundColor = .clear
        contentMode = .redraw
    }

    // MARK: Layout

    public override var intrinsicContentSize: CGSize {
        guard let r = renderer else { return .zero }
        return CGSize(width: r.width, height: r.totalHeight)
    }

    // MARK: Drawing

    public override func draw(_ rect: CGRect) {
        guard let renderer, let ctx = UIGraphicsGetCurrentContext() else { return }

        let contentW = renderer.width
        let contentH = renderer.totalHeight
        let availW = max(0, bounds.width)
        let availH = max(0, bounds.height)

        ctx.saveGState()
        ctx.clip(to: bounds)

        // Scale down to fit in the assigned layout size; never scale up.
        let sx: CGFloat = contentW > 0 ? (availW / contentW) : 1
        let sy: CGFloat = contentH > 0 ? (availH / contentH) : 1
        let scale = min(1, min(sx, sy))

        let scaledW = contentW * scale
        let scaledH = contentH * scale
        let dx = max(0, (availW - scaledW) / 2)
        let dy = max(0, (availH - scaledH) / 2)

        ctx.translateBy(x: dx, y: dy)
        ctx.scaleBy(x: scale, y: scale)
        renderer.draw(in: ctx)
        ctx.restoreGState()
    }

    public override func traitCollectionDidChange(_ previousTraitCollection: UITraitCollection?) {
        super.traitCollectionDidChange(previousTraitCollection)
        guard let previousTraitCollection else { return }
        guard traitCollection.hasDifferentColorAppearance(comparedTo: previousTraitCollection) else {
            return
        }
        rerender()
    }

    // MARK: Private

    private func rerender() {
        // Parsing + layout is < 1ms — run synchronously on the main thread.
        // Async dispatch would cause UITableView/List to lock in a zero height
        // before the render completes, making cells invisible.
        RaTeXFontLoader.ensureLoaded()
        do {
            let dl = try RaTeXEngine.shared.parse(
                latex,
                displayMode: displayMode,
                color: color,
                traitCollection: traitCollection
            )
            renderer = RaTeXRenderer(displayList: dl, fontSize: fontSize)
            invalidateIntrinsicContentSize()
            setNeedsDisplay()
        } catch {
            onError?(error)
        }
    }
}

// MARK: - SwiftUI

/// A SwiftUI view that renders a LaTeX formula.
///
/// ```swift
/// RaTeXFormula(latex: #"\int_0^\infty e^{-x^2}\,dx = \frac{\sqrt{\pi}}{2}"#, fontSize: 24)
/// ```
@available(iOS 14, *)
public struct RaTeXFormula: UIViewRepresentable {
    public let latex: String
    public var fontSize: CGFloat = 24
    public var onError: ((Error) -> Void)? = nil

    public init(latex: String, fontSize: CGFloat = 24, onError: ((Error) -> Void)? = nil) {
        self.latex = latex
        self.fontSize = fontSize
        self.onError = onError
    }

    public func makeUIView(context: Context) -> RaTeXView {
        let view = RaTeXView()
        view.setContentHuggingPriority(.required, for: .horizontal)
        view.setContentHuggingPriority(.required, for: .vertical)
        return view
    }

    public func updateUIView(_ uiView: RaTeXView, context: Context) {
        uiView.fontSize = fontSize
        uiView.onError  = onError
        uiView.latex    = latex
    }
}

File: platforms/react-native/android/src/main/kotlin/io/ratex/RaTeXView.kt
// RaTeXView.kt — Android custom View that renders a LaTeX formula.

package io.ratex

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.util.AttributeSet
import android.view.View
import androidx.annotation.ColorInt
import kotlin.math.max
import kotlin.math.min
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * A custom [View] that renders a LaTeX math formula using the RaTeX engine.
 *
 * XML usage:
 * ```xml
 * <io.ratex.RaTeXView
 *     android:id="@+id/mathView"
 *     android:layout_width="wrap_content"
 *     android:layout_height="wrap_content"
 *     app:latex="\frac{1}{2}"
 *     app:fontSize="24" />
 * ```
 *
 * Kotlin usage:
 * ```kotlin
 * binding.mathView.latex    = """\frac{-b \pm \sqrt{b^2-4ac}}{2a}"""
 * binding.mathView.fontSize = 28f   // dp — converted to px internally
 * ```
 */
class RaTeXView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0,
) : View(context, attrs, defStyle) {

    // MARK: - Public properties

    /** LaTeX math-mode string to render. Setting this triggers an async re-render. */
    var latex: String = ""
        set(value) {
            if (field == value) return
            field = value
            rerender()
        }

    /**
     * Font size in density-independent units (dp), matching React Native / iOS points.
     * Setting this triggers an async re-render.
     */
    var fontSize: Float = 24f
        set(value) {
            if (field == value) return
            field = value
            rerender()
        }

    /**
     * Rendering mode. `true` (default) for display/block style (`$$...$$`);
     * `false` for inline/text style (`$...$`). Setting this triggers an async re-render.
     */
    var displayMode: Boolean = true
        set(value) {
            if (field == value) return
            field = value
            rerender()
        }

    /** Default formula color. Explicit LaTeX colors still take precedence. */
    @ColorInt
    var color: Int = Color.BLACK
        set(value) {
            if (field == value) return
            field = value
            rerender()
        }

    /** Called on the main thread when a render error occurs. */
    var onError: ((RaTeXException) -> Unit)? = null

    /** Called on the main thread when content size is known (width/height in dp). */
    var onContentSizeChange: ((width: Double, height: Double) -> Unit)? = null

    // MARK: - Private state

    private var renderer: RaTeXRenderer? = null
    private val scope = CoroutineScope(Dispatchers.Main + SupervisorJob())
    private var renderJob: Job? = null

    // MARK: - Measure

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val r = renderer
        val desiredWidth = max(
            (r?.widthPx?.toInt() ?: 0) + paddingLeft + paddingRight,
            suggestedMinimumWidth,
        )
        val desiredHeight = max(
            (r?.totalHeightPx?.toInt() ?: 0) + paddingTop + paddingBottom,
            suggestedMinimumHeight,
        )

        // Respect parent / RN layout constraints (e.g. style={{width,height}}).
        val measuredWidth = resolveSize(desiredWidth, widthMeasureSpec)
        val measuredHeight = resolveSize(desiredHeight, heightMeasureSpec)
        setMeasuredDimension(measuredWidth, measuredHeight)
    }

    // MARK: - Draw

    override fun onDraw(canvas: Canvas) {
        val r = renderer ?: return

        val availW = (width - paddingLeft - paddingRight).toFloat().coerceAtLeast(0f)
        val availH = (height - paddingTop - paddingBottom).toFloat().coerceAtLeast(0f)
        val contentW = r.widthPx
        val contentH = r.totalHeightPx

        // Clip to the view bounds so explicit style sizes behave predictably.
        canvas.save()
        canvas.clipRect(0, 0, width, height)

        // Scale down to fit within the explicit layout size (never scale up).
        val sx = if (contentW > 0f) availW / contentW else 1f
        val sy = if (contentH > 0f) availH / contentH else 1f
        val scale = min(1f, min(sx, sy))

        val scaledW = contentW * scale
        val scaledH = contentH * scale

        val dx = paddingLeft + ((availW - scaledW) / 2f).coerceAtLeast(0f)
        val dy = paddingTop + ((availH - scaledH) / 2f).coerceAtLeast(0f)

        canvas.translate(dx, dy)
        canvas.scale(scale, scale)
        r.draw(canvas)
        canvas.restore()
    }

    // MARK: - Lifecycle

    override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        renderJob?.cancel()
        renderJob = null
    }

    // MARK: - Private

    private fun rerender() {
        renderJob?.cancel()
        if (latex.isBlank()) {
            renderer = null
            requestLayout()
            invalidate()
            return
        }
        renderJob = scope.launch {
            try {
                withContext(Dispatchers.IO) { RaTeXFontLoader.ensureLoaded(context) }
                val dl = RaTeXEngine.parse(latex, displayMode, color)
                // RN passes logical size (dp); convert to px so physical size matches iOS points.
                val density = context.resources.displayMetrics.density
                val fontSizePx = fontSize * density
                val r = RaTeXRenderer(dl, fontSizePx) { RaTeXFontLoader.getTypeface(it) }
                renderer = r
                requestLayout()
                invalidate()
                val widthDp = r.widthPx / density
                val heightDp = r.totalHeightPx / density
                onContentSizeChange?.invoke(widthDp.toDouble(), heightDp.toDouble())
            } catch (e: RaTeXException) {
                renderer = null
                requestLayout(); invalidate()
                onError?.invoke(e)
            } catch (e: Throwable) {
                renderer = null
                requestLayout(); invalidate()
                onError?.invoke(RaTeXException(e.message ?: "unknown error"))
            }
        }
    }
}
